cli/azd/internal/vsrpc/environment_service_refresh.go

```diff
@@ -0,0 +1,187 @@
+		return nil, err
+	}
+
+	infra, err := c.importManager.ProjectInfrastructure(ctx, c.projectConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = infra.Cleanup() }()
+
+	if err := bicepProvider.Initialize(ctx, c.projectConfig.Path, infra.Options); err != nil {
+		return nil, fmt.Errorf("initializing provisioning manager: %w", err)
+	}
+
+	_ = observer.OnNext(ctx, newInfoProgressMessage("Loading latest deployment information"))
+
+	deployment, err := bicepProvider.LastDeployment(ctx)
+	if err != nil {
+		log.Printf("failed to get latest deployment result: %v", err)
+	} else {
+		env.LastDeployment = &DeploymentResult{
+			DeploymentId: *deployment.ID,
+			Success:      *deployment.Properties.ProvisioningState == armresources.ProvisioningStateSucceeded,
+			Time:         *deployment.Properties.Timestamp,
+		}
+	}
+
+	stableServices, err := c.importManager.ServiceStable(ctx, c.projectConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	subId := *env.Properties["Subscription"]
+	envName := env.Name
+
+	nameIdx := make(map[string]int, len(env.Services)) // maps service name to index in env.Services slice
+	for idx, svc := range env.Services {
+		nameIdx[svc.Name] = idx
+	}
+
+	_ = observer.OnNext(ctx, newInfoProgressMessage("Loading server resources"))
+
+	rgName, err := c.azureResourceManager.FindResourceGroupForEnvironment(ctx, subId, envName)
+	if err == nil {
+		env.Properties["ResourceGroup"] = convert.RefOf(rgName)
+
+		for _, serviceConfig := range stableServices {
+			svcName := serviceConfig.Name
+
+			_ = observer.OnNext(ctx, newInfoProgressMessage("Loading server resources for service "+svcName))
+
+			resources, err := c.resourceManager.GetServiceResources(ctx, subId, rgName, serviceConfig)
+			if err == nil {
+				resourceIds := make([]string, len(resources))
+				for idx, res := range resources {
+					resourceIds[idx] = res.Id
+				}
+
+				if svcIdx, has := nameIdx[svcName]; has {
+					resSvc := env.Services[svcIdx]
+
+					if len(resourceIds) > 0 {
+						resSvc.ResourceId = convert.RefOf(resourceIds[0])
+					}
+
+					resSvc.Endpoint = convert.RefOf(s.serviceEndpoint(
+						ctx, subId, serviceConfig, c.resourceManager, c.serviceManager,
+					))
+				}
+			} else {
+				log.Printf("ignoring error determining resource id for service %s: %v", svcName, err)
+			}
+		}
+	} else {
+		log.Printf(
+			"ignoring error determining resource group for environment %s, resource ids will not be available: %v",
+			env.Name,
+			err)
+	}
+
+	return env, nil
+}
+
+func (s *environmentService) serviceEndpoint(
+	ctx context.Context,
+	subId string,
+	serviceConfig *project.ServiceConfig,
+	resourceManager project.ResourceManager,
+	serviceManager project.ServiceManager,
+) string {
+	targetResource, err := resourceManager.GetTargetResource(ctx, subId, serviceConfig)
+	if err != nil {
+		log.Printf("error: getting target-resource. Endpoints will be empty: %v", err)
```

Comment:
These things do fail in-practice in azd when ARM 404s on newly deployed resources.. Is there an ideal way we would surface this as warning messages to VS? Could be a separate issue around how reporting would work here.